File: stream_processor.py
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, from_json, to_json, struct, udf
from pyspark.sql.types import StructType, StructField, StringType, DoubleType
from pyspark.ml import PipelineModel

def main():
    """
    Main function to run the Spark Structured Streaming application.
    """
    spark = SparkSession.builder \
        .appName("RealTimeFraudDetection") \
        .config("spark.jars.packages", "org.apache.spark:spark-sql-kafka-0-10_2.12:3.3.0") \
        .getOrCreate()

    spark.sparkContext.setLogLevel("WARN")

    print("--- Starting Real-Time Fraud Detection Stream ---")

    # --- 1. Load the pre-trained ML model ---
    model_path = "medium-stories/data-enhancement/spark_ml_model"
    print(f"Loading pre-trained model from {model_path}...")
    model = PipelineModel.load(model_path)
    print("Model loaded successfully.")

    # --- 2. Define the schema for incoming Kafka data ---
    # This must match the structure of the JSON data produced by producer.py
    kafka_schema = StructType([
        StructField("type", StringType(), True),
        StructField("amount", DoubleType(), True),
        StructField("oldbalanceOrg", DoubleType(), True),
        StructField("newbalanceOrig", DoubleType(), True),
    ])

    # --- 3. Read from the 'raw_transactions' Kafka topic ---
    kafka_bootstrap_servers = 'kafka-cluster.local:9092'
    raw_transactions_topic = 'raw_transactions'

    print(f"Subscribing to Kafka topic '{raw_transactions_topic}'...")
    kafka_stream_df = spark \
        .readStream \
        .format("kafka") \
        .option("kafka.bootstrap.servers", kafka_bootstrap_servers) \
        .option("subscribe", raw_transactions_topic) \
        .option("startingOffsets", "latest") \
        .load()

    # --- 4. Parse the JSON data and apply the ML model ---
    parsed_stream_df = kafka_stream_df \
        .select(from_json(col("value").cast("string"), kafka_schema).alias("data")) \
        .select("data.*")

    print("Applying ML model to the stream...")
    predictions_df = model.transform(parsed_stream_df)

    # --- 5. Format the output for the 'enhanced_transactions' topic ---
    extract_prob_udf = udf(lambda v: float(v[1]), DoubleType())

    output_df = predictions_df.withColumn("fraud_probability", extract_prob_udf(col("probability")))

    final_df = output_df.select(
        to_json(
            struct(
                col("type"),
                col("amount"),
                col("oldbalanceOrg"),
                col("newbalanceOrig"),
                col("prediction").alias("fraud_prediction"),
                col("fraud_probability")
            )
        ).alias("value")
    )

    # --- 6. Write the enhanced data to the 'enhanced_transactions' Kafka topic ---
    enhanced_transactions_topic = 'enhanced_transactions'
    checkpoint_location = '/tmp/spark_checkpoints/fraud_detection'
    
    print(f"Writing enhanced stream to Kafka topic '{enhanced_transactions_topic}'...")
    query = final_df \
        .writeStream \
        .format("kafka") \
        .option("kafka.bootstrap.servers", kafka_bootstrap_servers) \
        .option("topic", enhanced_transactions_topic) \
        .option("checkpointLocation", checkpoint_location) \
        .start()

    print("Stream is now running. Waiting for termination...")
    print("Check the output of the consumer script to see the enhanced data.")
    query.awaitTermination()

if __name__ == "__main__":
    main() 

File: src/main/scala/StreamProcessor.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._
import org.apache.spark.ml.PipelineModel
import org.apache.spark.ml.linalg.Vector

object StreamProcessor {

  def main(args: Array[String]): Unit = {
    val spark: SparkSession = createSparkSession

    println("--- Starting Real-Time Fraud Detection Stream ---")

    val model: PipelineModel = loadModel

    val kafkaSchema: StructType = createSchema

    val kafkaBootstrapServers = "kafka-cluster.local:9092"
    val rawTransactionsTopic = "raw_transactions"

    val parsedStreamDF = spark
      .readStream
      .format("kafka")
      .option("kafka.bootstrap.servers", kafkaBootstrapServers)
      .option("subscribe", rawTransactionsTopic)
      .option("startingOffsets", "latest")
      .load()
      .select(from_json(col("value").cast("string"), kafkaSchema).alias("data"))
      .select("data.*")

    // --- Perform same feature engineering as in training ---
    println("Applying feature engineering to the stream...")
    val engineeredStreamDF = parsedStreamDF
      .withColumn("fromTo",
        when(col("nameOrig").contains("C") && col("nameDest").contains("C"), "CC")
          .when(col("nameOrig").contains("C") && col("nameDest").contains("M"), "CM")
          .when(col("nameOrig").contains("M") && col("nameDest").contains("C"), "MC")
          .when(col("nameOrig").contains("M") && col("nameDest").contains("M"), "MM")
          .otherwise(lit(null))
      )
      .filter(col("fromTo").isNotNull)
      .withColumn("HourOfDay", col("step") % 24)

    println("Applying ML model to the stream...")
    val predictionsDF = model.transform(engineeredStreamDF)

    // --- Format the output for the 'enhanced_transactions' topic ---
    val extractProbability = udf((v: Vector) => v(1))

    val outputDF = predictionsDF
      .withColumn("fraud_probability", extractProbability(col("probability")))
      .select(
        to_json(
          struct(
            col("type"),
            col("amount"),
            col("prediction").alias("fraud_prediction"),
            col("fraud_probability")
          )
        ).alias("value")
      )

    val enhancedTransactionsTopic = "enhanced_transactions"
    val checkpointLocation = "/tmp/spark_checkpoints/fraud_detection_scala"
    
    outputDF
      .writeStream
      .format("kafka")
      .option("kafka.bootstrap.servers", kafkaBootstrapServers)
      .option("topic", enhancedTransactionsTopic)
      .option("checkpointLocation", checkpointLocation)
      .start()
      .awaitTermination()
  }

  private def createSchema: StructType = {
    StructType(Array(
      StructField("step", IntegerType, nullable = true),
      StructField("type", StringType, nullable = true),
      StructField("amount", DoubleType, nullable = true),
      StructField("nameOrig", StringType, nullable = true),
      StructField("oldbalanceOrg", DoubleType, nullable = true),
      StructField("newbalanceOrig", DoubleType, nullable = true),
      StructField("nameDest", StringType, nullable = true),
      StructField("oldbalanceDest", DoubleType, nullable = true),
      StructField("newbalanceDest", DoubleType, nullable = true),
      StructField("isFlaggedFraud", IntegerType, nullable = true)
    ))
  }

  private def loadModel: PipelineModel = {
    val modelPath = "/home/damian/business/medium-stories/data-enhancement/data-enhancement-model/spark_ml_model"
    val model = PipelineModel.load(modelPath)
    println("Model loaded successfully.")
    model
  }

  private def createSparkSession = {
    val spark = SparkSession.builder
      .appName("RealTimeFraudDetection")
      .master("local[*]")
      .getOrCreate()
    spark.sparkContext.setLogLevel("WARN")
    spark
  }
}